src/streams.h

```diff
@@ -15,32 +15,40 @@
 #include <assert.h>
 #include <cstddef>
 #include <cstdio>
+#include <cstring>
 #include <ios>
 #include <limits>
 #include <optional>
+#include <span>
 #include <stdint.h>
 #include <string.h>
 #include <string>
-#include <utility>
 #include <vector>
 
 namespace util {
 inline void Xor(Span<std::byte> write, Span<const std::byte> key, size_t key_offset = 0)
 {
-    if (key.size() == 0) {
-        return;
+    if (write.empty() || key.empty()) return;
+
+    size_t i = 0;
+    // Consume excess key_offset prefix
+    for (key_offset %= key.size(); key_offset > 0 && i < std::min(key.size() - key_offset, write.size()); ++i) {
+        write[i] ^= key[key_offset + i];
+    }
+
+    if (key.size() == 8 && write.size() - i >= 8) { // Xor in 64-bit chunks
+        uint64_t key64;
+        std::memcpy(&key64, key.data(), 8);
+        for (; i <= write.size() - 8; i += 8) {
+            uint64_t write64;
+            std::memcpy(&write64, write.data() + i, 8);
```

Comment:
i have a hard time believing this will make a large difference, especially with the two memcpys involved
On modern CPUs, ALU operations (especially bitwise ones) are so fast compared to any kind of memory access.
And this isn't some advanced crypto math, it's *one* Xor operation per word with a fixed key.

Could avoid the memcpys if the code takes memory alignment into account, but that makes it even more complex. Not sure the pros/cons work out here.

Reply:
The speedup comes from the vectorized operations, i.e. doing 64 bit xor instead of byte-by-byte xor (memcpy seems to be eliminated on 64 bit architectures successfully), see https://godbolt.org/z/Koscjconz

Reply:
Added RISC-V compiler to https://godbolt.org/z/n5rMeYeas - where it seems to my untrained eyes that it uses two separate 32 bit xors to emulate the 64 bit operation:
```asm
xor     a4,a4,s0
xor     a5,a5,t2
srli    t6,a4,8
srli    t5,a4,16
srli    t4,a4,24
srli    a7,a5,8
```

Reply:
I've replaced all vectors to 64 bit ints, the new peproducer is: [godbolt.org/z/sc5xPTcW6](https://godbolt.org/z/sc5xPTcW6), which shows:
* _Clang_ (x86-64) - uses SSE (`movdqu`/`pshufd`/`pxor`) for SIMD operations, processing **32** bytes / iteration
* _GCC_ (x86-64) -  uses simple 64-bit operations (`xor` with `QWORD PTR`), processes **16** bytes / iteration
* _RISC-V_ (32-bit) - uses basic 32-bit register operations (`lbu`, `xor`, `srli`, `sb`), processes **8** bytes / iteration

(please validate, my assembly knowledge is only academic)